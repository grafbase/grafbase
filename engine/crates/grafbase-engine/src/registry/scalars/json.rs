use grafbase_engine_value::ConstValue;
use serde::Deserialize;

use super::{DynamicParse, SDLDefinitionScalar};
use crate::{Error, InputValueError, InputValueResult};

pub struct JSONScalar;

impl<'a> SDLDefinitionScalar<'a> for JSONScalar {
    fn name() -> Option<&'a str> {
        Some("JSON")
    }

    fn description() -> Option<&'a str> {
        Some("A JSON Value")
    }
}

impl DynamicParse for JSONScalar {
    fn is_valid(_: &ConstValue) -> bool {
        true
    }

    fn to_value(value: serde_json::Value) -> Result<ConstValue, Error> {
        ConstValue::deserialize(value).map_err(|error| Error::new(error.to_string()))
    }

    fn parse(value: ConstValue) -> InputValueResult<serde_json::Value> {
        serde_json::Value::deserialize(value).map_err(|error| InputValueError::ty_custom("JSON", error.to_string()))
    }
}

#[cfg(test)]
mod tests {
    use grafbase_engine_value::ConstValue;
    use insta::assert_snapshot;

    use super::super::SDLDefinitionScalar;
    use crate::registry::scalars::{DynamicParse, JSONScalar};

    #[test]
    fn check_json_valid() {
        let value = serde_json::json!({
            "code": 200,
            "success": true,
            "payload": {
                "features": [
                    "serde",
                    "json"
                ]
            }
        });

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = JSONScalar::parse(const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn ensure_directives_sdl() {
        assert_snapshot!(JSONScalar::sdl());
    }
}
