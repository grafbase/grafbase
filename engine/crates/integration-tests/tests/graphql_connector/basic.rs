use integration_tests::{runtime, EngineBuilder, MockGraphQlServer, ResponseExt};
use serde_json::json;

const NAMESPACED_QUERY: &str = "
    query($id: ID!) {
        gothub {
            serverVersion
            pullRequestOrIssue(id: $id) {
                __typename
                title
                ... on GothubPullRequest {
                    checks
                    author {
                        __typename
                        ...AuthorFragment
                    }
                }
                ... on GothubIssue {
                    title
                    author {
                        __typename
                        ...AuthorFragment
                    }
                }
            }
        }
    }

    fragment AuthorFragment on GothubUserOrBot {
        ... on GothubUser {
            email
        }
        ... on GothubBot {
            id
        }
    }
";

#[test]
fn graphql_test_with_namespace() {
    runtime().block_on(async {
        let graphql_mock = MockGraphQlServer::new().await;

        let engine = EngineBuilder::new(schema(graphql_mock.port(), true)).build().await;

        insta::assert_json_snapshot!(
            "namespaced-pull-request-with-user",
            engine
                .execute(NAMESPACED_QUERY)
                .variables(json!({"id": "1"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "namespaced-pull-request-with-bot",
            engine
                .execute(NAMESPACED_QUERY)
                .variables(json!({"id": "2"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "namespaced-issue",
            engine
                .execute(NAMESPACED_QUERY)
                .variables(json!({"id": "3"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "namespaced-null",
            engine
                .execute(NAMESPACED_QUERY)
                .variables(json!({"id": "4"}))
                .await
                .into_value()
        );
    })
}

const UNNAMESPACED_QUERY: &str = "
    query($id: ID!) {
        serverVersion
        pullRequestOrIssue(id: $id) {
            __typename
            title
            ... on PullRequest {
                checks
                author {
                    __typename
                    ...AuthorFragment
                }
            }
            ... on Issue {
                title
                author {
                    __typename
                    ...AuthorFragment
                }
            }
        }
    }

    fragment AuthorFragment on UserOrBot {
        ... on User {
            email
        }
        ... on Bot {
            id
        }
    }
";

#[test]
fn graphql_test_without_namespace() {
    runtime().block_on(async {
        let graphql_mock = MockGraphQlServer::new().await;

        let engine = EngineBuilder::new(schema(graphql_mock.port(), false)).build().await;

        insta::assert_json_snapshot!(
            "unnamespaced-pull-request-with-user",
            engine
                .execute(UNNAMESPACED_QUERY)
                .variables(json!({"id": "1"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "unnamespaced-pull-request-with-bot",
            engine
                .execute(UNNAMESPACED_QUERY)
                .variables(json!({"id": "2"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "unnamespaced-issue",
            engine
                .execute(UNNAMESPACED_QUERY)
                .variables(json!({"id": "3"}))
                .await
                .into_value()
        );
        insta::assert_json_snapshot!(
            "unnamespaced-null",
            engine
                .execute(UNNAMESPACED_QUERY)
                .variables(json!({"id": "4"}))
                .await
                .into_value()
        );
    });
}

fn schema(port: u16, namespace: bool) -> String {
    format!(
        r#"
          extend schema
          @graphql(
            name: "gothub",
            namespace: {namespace},
            url: "http://127.0.0.1:{port}",
            schema: "http://127.0.0.1:{port}/spec.json",
          )
        "#
    )
}
