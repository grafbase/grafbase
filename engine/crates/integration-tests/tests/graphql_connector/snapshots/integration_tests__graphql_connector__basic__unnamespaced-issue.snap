---
source: crates/integration-tests/tests/graphql_connector/basic.rs
expression: "engine.execute(UNNAMESPACED_QUERY).variables(json!({\n                    \"id\" : \"3\"\n                })).await.into_value()"
---
{
  "data": {
    "pullRequestOrIssue": {
      "__typename": "Issue",
      "author": {
        "__typename": "User",
        "email": "pessimist@example.com"
      },
      "title": "Everythings fucked"
    },
    "serverVersion": "1"
  }
}
